File: packages/color/src/entities/color-group.ts
import { Color } from "./color";

class ColorGroup {
  public main: Color = new Color(0, 0, 0, 0);
  public num: number = 0;

  public setMainColor(color: Color) {
    this.main = color;
  }

  public getDistance(c: Color) {
    return Math.abs(this.main.r - c.r) + Math.abs(this.main.g - c.g) + Math.abs(this.main.b - c.b);
  }

  public addColor(color: Color) {
    this.main.r = Math.floor((this.main.r + color.r) / 2);
    this.main.g = Math.floor((this.main.g + color.g) / 2);
    this.main.b = Math.floor((this.main.b + color.b) / 2);

    this.num += 1;
  }
}

export { ColorGroup };

File: crates/kit-k-means/src/color_group.rs
use crate::color::Color;
use js_sys::Math;

pub struct ColorGroup {
  pub main: Color,
  pub num: u32,
}

impl ColorGroup {
  pub fn new() -> ColorGroup {
    ColorGroup {
      main: Color::new(0.0, 0.0, 0.0, 0.0),
      num: 0,
    }
  }

  pub fn set_main_color(&mut self, color: &Color) {
    self.main = *color;
  }

  pub fn get_distance(&self, c: Color) -> f64 {
    return Math::abs((self.main.r - c.r).into())
      + Math::abs((self.main.g - c.g).into())
      + Math::abs((self.main.b - c.b).into());
  }

  pub fn add_color(&mut self, color: Color) {
    self.num = self.num + 1;

    self.main.r = (self.main.r + color.r) / 2.0;
    self.main.g = (self.main.g + color.g) / 2.0;
    self.main.b = (self.main.b + color.b) / 2.0;
  }
}
